### ISSUE TITLE
DafnyServer no longer caches results

### ISSUE THREAD

**Issue by wilcoxjay** [Microsoft/dafny#123]
I believe this was introduced by #114. Was the behavior of disabling caching intentional, though? Based on the discussion there, it looks like it may not have been intentional. 

Despite the known whitespace issues with caching, I still find its benefits far outweigh its costs. Perhaps a command line flag could be added to the server to enable/disable caching (does such a flag already exist?), so that the vscode plugin can turn it off.

**Comment by cpitclaudel** [comment]
I find this annoying too, and I'm surprised by @markusschaden's silence on the topic…  It's disheartening to invest time in building a Dafny server (and documenting the resulting protocol) with an explicit focus on making it easy to add Dafny support to other IDEs, only to find subsequent changes accepted with no prior discussion, no documentation, and apparently little concern about backwards compatibility.

I'm sure this is just a misunderstanding — documenting the new commands and restoring the old caching behavior could go a long way in clearing it. Ideally, future extensions can then be publicly discussed before they are implemented and merged.

**Comment by markusschaden** [comment]
Sorry for the late response, I was off for two weeks and did not had the time to response.
Adding a command line flag would be definitely a good idea. 
Should I just make a new pullrequest to enable caching again?

I tried to add all new features to the documentation, but didn't find the right spot. Have a look at the attached PDF, chapter 2.2. All the introduced verbs are explained with an example. 
[DafnyAPI.pdf](https://github.com/Microsoft/dafny/files/1100554/DafnyAPI.pdf)

PS: @misto have a look at this discussion and [pullrequest 114](https://github.com/Microsoft/dafny/pull/114)

**Comment by wilcoxjay** [comment]
A few weeks ago, Rustan added better error location rewriting to Boogie. I have been using it under a (caching-enabled) Dafny server since then and have yet to run into any issues with bad locations.

If anyone else is interested in testing this, ensure you have a recent Boogie (and Dafny), and look at  
[Snapshots9.v0.dfy](https://github.com/Microsoft/dafny/blob/master/Test/dafny0/snapshots/Inputs/Snapshots9.v0.dfy). After it loads in your editor, edit the contents (adding whitespace) to make it look like [Snapshots9.v1.dfy](https://github.com/Microsoft/dafny/blob/master/Test/dafny0/snapshots/Inputs/Snapshots9.v1.dfy). Before the recent changes, some of the errors would not be moved. Now they are all moved correctly.

So my new proposal is to just re-enable caching (at level 3, which uses the error location rewriting mechanism) and not bother to add a command line flag to disable it in the server.

@markusschaden Do you have any objection to this approach?

**Comment by markusschaden** [comment]
@wilcoxjay Adding a command line flag sounds promising. So it is possible to use the server, with or without caching enabled. Otherwise you can also revert my change and just enable caching again.

**Comment by wilcoxjay** [comment]
I have just pushed [b7c3b33](https://github.com/Microsoft/dafny/commit/b7c3b333431182e4fc00d429b35cf547e45c21ef), which re-enables caching. 

I set the option in a way that permits users of the server to override it on a per-request basis. In other words, in order to disable caching, the option /verifySnapshots:0 needs to be passed with each verify request.

If there is sufficient interest, I can look into making it possible to pass this option only once, at server-start time.
